File: server/nodejs/src/services/userService.js
const fs = require('fs');
const path = require('path');
const { v4: uuidv4 } = require('uuid');
const bcrypt = require('bcrypt');
const dbConfig = require('../config/dbConfig');
const userSchema = require('../models/userSchema');

const dbPath = path.join(__dirname, `../../data/${dbConfig.databaseUrl}/users.json`);

// Ensure the data directory exists
fs.mkdirSync(path.dirname(dbPath), { recursive: true });

// Initialize an empty JSON file if it doesn't exist
if (!fs.existsSync(dbPath)) {
    fs.writeFileSync(dbPath, JSON.stringify([]));
}

function readUsers() {
    return JSON.parse(fs.readFileSync(dbPath));
}

function writeUsers(users) {
    fs.writeFileSync(dbPath, JSON.stringify(users));
}

async function createUser(user) {
    const users = readUsers();

    const existingUser = users.find(u => u.email === user.email);
    if (existingUser) {
        throw new Error('User with this email already exists');
    }

    user.id = uuidv4();
    user.password = await bcrypt.hash(user.password, 10);
    const { error } = userSchema.validate(user);
    if (error) throw error;

    users.push(user);
    writeUsers(users);
    return user;
}

function getUser(id) {
    const users = readUsers();
    return users.find(user => user.id === id);
}

function updateUser(id, updates) {
    const users = readUsers();
    const userIndex = users.findIndex(user => user.id === id);

    if (userIndex === -1) {
        throw new Error('User not found');
    }

    const updatedUser = { ...users[userIndex], ...updates, updated_at: new Date() };
    const { error } = userSchema.validate(updatedUser);
    if (error) throw error;

    users[userIndex] = updatedUser;
    writeUsers(users);
    return updatedUser;
}

function deleteUser(id) {
    const users = readUsers();
    const userIndex = users.findIndex(user => user.id === id);

    if (userIndex === -1) {
        throw new Error('User not found');
    }

    const [deletedUser] = users.splice(userIndex, 1);
    writeUsers(users);
    return deletedUser;
}

async function authenticateUser(email, password) {
    const users = readUsers();
    const user = users.find(user => user.email === email);

    if (!user || !(await bcrypt.compare(password, user.password))) {
        throw new Error('Invalid email or password');
    }

    return user;
}

module.exports = { createUser, getUser, updateUser, deleteUser, authenticateUser };

File: server/python/src/services/user_service.py
import json
import os
from models.user import User

class UserService:
    def __init__(self, db_path):
        self.db_path = db_path
        if not os.path.exists(db_path):
            with open(db_path, 'w') as f:
                json.dump([], f)

    def read_users(self):
        with open(self.db_path, 'r') as f:
            return [User.from_dict(user) for user in json.load(f)]

    def write_users(self, users):
        with open(self.db_path, 'w') as f:
            json.dump([user.to_dict() for user in users], f, indent=4)

    def create_user(self, user):
        users = self.read_users()
        if any(u.email == user.email for u in users):
            raise ValueError('User with this email already exists')
        users.append(user)
        self.write_users(users)
        return user

    def get_user(self, id):
        users = self.read_users()
        for user in users:
            if user.id == id:
                return user
        return None

    def update_user(self, id, updates):
        users = self.read_users()
        for i, user in enumerate(users):
            if user.id == id:
                user.username = updates.username or user.username
                user.email = updates.email or user.email
                user.password = updates.password or user.password
                user.updated_at = datetime.now().isoformat()
                users[i] = user
                self.write_users(users)
                return user
        raise ValueError('User not found')

    def delete_user(self, id):
        users = self.read_users()
        users = [user for user in users if user.id != id]
        self.write_users(users)
        return True

    def authenticate_user(self, email, password):
        users = self.read_users()
        for user in users:
            if user.email == email and user.password == User.hash_password(password):
                return user
        return None
